starlark/eval.go

```diff
@@ -478,26 +478,18 @@ func setIndex(fr *Frame, x, y, z Value) error {
 
 // Unary applies a unary operator (+, -, ~, not) to its operand.
 func Unary(op syntax.Token, x Value) (Value, error) {
-	switch op {
-	case syntax.MINUS:
-		switch x := x.(type) {
-		case Int:
-			return zero.Sub(x), nil
-		case Float:
-			return -x, nil
-		}
-	case syntax.PLUS:
-		switch x.(type) {
-		case Int, Float:
-			return x, nil
-		}
-	case syntax.TILDE:
-		if xint, ok := x.(Int); ok {
-			return xint.Not(), nil
-		}
-	case syntax.NOT:
+	if op == syntax.NOT {
 		return !x.Truth(), nil
 	}
+
+	// Int, Float, and user-defined types
+	if x, ok := x.(HasUnary); ok {
```

Comment:
No change needed in this PR, but does it make sense to move logic from `Binary` to `HasBinary` implementations for builtin types? I can't remember if there was a good reason to have everything in `Binary` or if `HasBinary` simply came later.

Reply:
It seems tempting to do so, but it can't be done in languages without multimethod dispatch. This is a classic problem in object-oriented languages called the "binary method problem"; see for example
https://lib.dr.iastate.edu/cgi/viewcontent.cgi?article=1032&context=cs_techreports